lib/lz4.c

```diff
@@ -297,6 +297,75 @@ void LZ4_wildCopy(void* dstPtr, const void* srcPtr, void* dstEnd)
     do { memcpy(d,s,8); d+=8; s+=8; } while (d<e);
 }
 
+static const unsigned inc32table[8] = {0, 1, 2,  1,  0,  4, 4, 4};
+static const int      dec64table[8] = {0, 0, 0, -1, -4,  1, 2, 3};
+
+
+LZ4_FORCE_O2_INLINE_GCC_PPC64LE
+void LZ4_memcpy_using_offset_base(BYTE* dstPtr, const BYTE* srcPtr, BYTE* dstEnd, const size_t offset) {
+    if (offset < 8) {
+        dstPtr[0] = srcPtr[0];
+        dstPtr[1] = srcPtr[1];
+        dstPtr[2] = srcPtr[2];
+        dstPtr[3] = srcPtr[3];
+        srcPtr += inc32table[offset];
+        memcpy(dstPtr+4, srcPtr, 4);
+        srcPtr -= dec64table[offset];
+        dstPtr += 8;
+    } else {
+        memcpy(dstPtr, srcPtr, 8);
+        dstPtr += 8;
+        srcPtr += 8;
+    }
+
+    LZ4_wildCopy(dstPtr, srcPtr, dstEnd);
+}
+
+/* customized variant of memcpy, which can overwrite up to 32 bytes beyond dstEnd */
+LZ4_FORCE_O2_INLINE_GCC_PPC64LE
+void LZ4_wildCopy32(void* dstPtr, const void* srcPtr, void* dstEnd)
+{
+    BYTE* d = (BYTE*)dstPtr;
+    const BYTE* s = (const BYTE*)srcPtr;
+    BYTE* const e = (BYTE*)dstEnd;
+
+    memcpy(d,s,16); d+=16; s+=16;
```

Comment:
question : 
this function selects to copy 16 bytes first,
and then, provided that length is large enough, will proceed with further stripes of 32 bytes.

Any reason it starts with 16 bytes, instead of 32 bytes ?
I guess you made tests to reach this conclusion, and maybe could share some results ?

Reply:
Starting with 16, then doing a copy loop of 32:

```
lz4 1.8.3                 319 MB/s  2914 MB/s     6428742  63.07 silesia/dickens
lz4 1.8.3                 493 MB/s  2614 MB/s    26435667  51.61 silesia/mozilla
lz4 1.8.3                 494 MB/s  3376 MB/s     5440937  54.57 silesia/mr
lz4 1.8.3                 845 MB/s  3884 MB/s     5533040  16.49 silesia/nci
lz4 1.8.3                 412 MB/s  2651 MB/s     4338918  70.53 silesia/ooffice
lz4 1.8.3                 456 MB/s  2977 MB/s     5256666  52.12 silesia/osdb
lz4 1.8.3                 329 MB/s  2483 MB/s     3181387  48.00 silesia/reymont
lz4 1.8.3                 557 MB/s  3516 MB/s     7716839  35.72 silesia/samba
lz4 1.8.3                 441 MB/s  3985 MB/s     6790273  93.63 silesia/sao
lz4 1.8.3                 361 MB/s  2714 MB/s    20139988  48.58 silesia/webster
lz4 1.8.3                1218 MB/s  8687 MB/s     8390195  99.01 silesia/x-ray
lz4 1.8.3                 718 MB/s  3611 MB/s     1227495  22.96 silesia/xml
```

Going straight to a 32byte copy loop:
```
lz4 1.8.3                 312 MB/s  3091 MB/s     6428742  63.07 silesia/dickens
lz4 1.8.3                 499 MB/s  2744 MB/s    26435667  51.61 silesia/mozilla
lz4 1.8.3                 496 MB/s  3486 MB/s     5440937  54.57 silesia/mr
lz4 1.8.3                 877 MB/s  4328 MB/s     5533040  16.49 silesia/nci
lz4 1.8.3                 424 MB/s  2839 MB/s     4338918  70.53 silesia/ooffice
lz4 1.8.3                 459 MB/s  3202 MB/s     5256666  52.12 silesia/osdb
lz4 1.8.3                 326 MB/s  2619 MB/s     3181387  48.00 silesia/reymont
lz4 1.8.3                 551 MB/s  3555 MB/s     7716839  35.72 silesia/samba
lz4 1.8.3                 431 MB/s  4116 MB/s     6790273  93.63 silesia/sao
lz4 1.8.3                 355 MB/s  2872 MB/s    20139988  48.58 silesia/webster
lz4 1.8.3                1206 MB/s  9369 MB/s     8390195  99.01 silesia/x-ray
lz4 1.8.3                 682 MB/s  3623 MB/s     1227495  22.96 silesia/xml
```

So yea we should change it to 32-byte, which makes sense: We only use this wildcopy when we are already copying >= 15 bytes.